text/0000-placement-by-return.md

```diff
@@ -0,0 +1,1097 @@
+
+* **It needs to be possible to wrap it in a safe API.** Safe API examples are given for built-in data structures, including a full sketch of the implementation for Box, including exception safety.
+* **It needs to support already-idiomatic constructors like `fn new() -> GiantStruct { GiantStruct { ... } }`** Since this proposal is defined in terms of Guaranteed Copy Elision, this is a gimme.
+* **It needs to be possible to in-place populate data structures that cannot be written using a single literal expression.** The `write_return_with` intrinsic allows this to be done in an unsafe way. Sketches for APIs built on top of them are also given in the [future-possibilities] section.
+* **It needs to avoid adding egregious overhead in cases where the values being populated are small (in other words, if the value being initialized is the size of a pointer or smaller, it needs to be possible for the compiler to optimize away the outptr).** Since this proposal is written in terms of Guaranteed Copy Elision, this is a gimme. The exception of the "weird bypass functions" `read_return_with` and `write_return_with` may seem to break this; see the [example desugarings here](#How-do-the-return-slot-functions-work-when-the-copy-is-not-actually-elided) for info on how these functions work when the copy is not actually elided.
+* **It needs to solve most of the listed problems with the old proposals.** Since this one actually goes the distance and defines when copy elision will kick in, it fixes the biggest problems that the old `box` literal system had. It is also written in terms of present-day Rust, using `impl Trait`, and with the current direction of Rust in mind.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If you need to allocate a very large data structure, or a DST, you should prefer using `Box::new_with` over `Box::new`. For example:
+
+```rust
+let boxed_array = Box::new_with(|| [0; 1_000_000]); // instead of Box::new([0; 1_000_000])
+let boxed_data_struct = Box::new_with(DataStruct::new); // instead of Box::new(DataStruct::new())
+```
+
+The `new_with` function will perform the allocation first, then evaluate the closure, placing the result directly within it. The `new` function, on the other hand, evaluates the argument *then* performs the allocation and copies the value into it.
+
+Similar functions exist for Vec and other data structures, with names like `push_with` and `insert_with`.
+
+When writing constructors, you should create large data structures directly on the return line. For example:
+
+```rust
+// This function will allocate its return value directly in the return slot.
+fn good() -> [i32; 1_000_000] {
+    [1; 1_000_000]
+}
+// This function will return a raw slice, and can also be called as `Box::new_with(good_dst)`
+fn good_dst() -> [i32] {
+    let n = 1_000_000;
+    [1; n]
+}
+// This function will copy the array when it returns.
```

Comment:
I would clarify this comment to say that while this could elide the copy, this is not guaranteed by the RFC for ease of implementation. Basically what you said in comments-- we don't support anything that isn't directly in the return slot for the initial pass.

Similar thing-- would `let arr = [0; 1_000_000]; arr` be copied as well, or are lets without mutation in the middle okay?

Reply:
Never mind, this is covered by a later example. Maybe also put that example up here?